### PR TITLE
feat: add `Decidable`, `Fintype`, `Encodable` linters

### PR DESCRIPTION
**Pull request by urkud**
This PR ports the `decidableClassical`, `fintypeFinite` and `encodableCountable` linters from mathlib3: all three are environment linters which flag declarations that have a hypothesis `[Decidable p]` etc., but do not use the underlying assumption in this type. Restructuring the code can make the statement more general.
The implementation is in fact quite general, and can easily adapt to include further types. PRs #17518 and #17519 are two examples of further such linters.

---

Please don't fix the failures related to `HasFDerivWithinAt.pi`. Finiteness assumptions in these lemmas will be completely removed during porting to TVS.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grunweg** [comment]
FYI: #15278, #15279 and #15280 fix some easy linter errors: I don't have further fixes in the pipeline

**Comment by urkud** [comment]
We should skip constructors of inductive types as well (at least, for now).

**Comment by grunweg** [comment]
Do you know how to do this? (You can also leave a pointer here and I'll try to add this to the linter, when I find time. This week, and probably the next, I'm also rather busy with moving places, so cannot make promises about my available Lean time.)

**Comment by grunweg** [comment]
!bench

**Comment by grunweg** [comment]
(I worry if the linter implementation is too slow right now, and the linters should rather be merged... let's benchmark to find out!)

**Review by kmill — APPROVED** [review]

File: Mathlib/GroupTheory/Perm/DomMulAct.lean [hunk 1]
```diff
@@ -106,6 +107,7 @@ theorem stabilizer_card [DecidableEq α] [DecidableEq ι] [Fintype ι] :
   · rfl
 
 /-- The cardinality of the set of permutations preserving a function -/
+@[nolint finiteFintype]
```
Comment (kmill) [inline Mathlib/GroupTheory/Perm/DomMulAct.lean:110]: What is the reason that this nolint is necessary? The `Fintype` instance is used directly in the `∏` notation.

Comment (grunweg) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean:110]: Good question, I'm not sure. @urkud Do you know?

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean [hunk 2]
```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
```
Comment (urkud) [inline Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:4]: @grunweg Do you want to add yourself here?

File: Mathlib/GroupTheory/Perm/DomMulAct.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean:110]: You don't need `[Fintype α]`, only `[Finite α]` here. In fact, I'm not sure if you need `[Finite α]` at all: if `α` is infinite, then one of the fibers is infinite, and RHS = 0. OTOH, LHS is infinite or empty, so you don't need `[Finite α]`.

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:4]: Thanks, that sounds reasonable.
```suggestion
Authors: Yury Kudryashov, Michael Rothgang

File: Mathlib/GroupTheory/Perm/DomMulAct.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean:110]: @urkud Do you have a moment to try to make this change? I won't, until next week.

**Comment by grunweg** [comment]
!bench

File: Mathlib/GroupTheory/Perm/DomMulAct.lean (hunk 1 above)
Comment (urkud) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean]: Fixed.

Comment (grunweg) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean]: Thanks!

**Comment by grunweg** [comment]
!bench

**Comment by grunweg** [comment]
@kmill Thanks for commenting about the `GeneralizeProofs` change; I'm learning something new every day!
The linter ignores `partial` declarations now; I just verified that the change *could* be reverted. Would you prefer that change?

**Comment by urkud** [comment]
@grunweg Linter flags `GeneralizeProofs` now. I can't find the relevant @kmill's comment. Was it here or on Zulip?

**Comment by grunweg** [comment]
It was [on a dependent PR](https://github.com/leanprover-community/mathlib4/pull/17125#issuecomment-2405497324).

**Comment by grunweg** [comment]
I'm also open to reverting [this commit](https://github.com/leanprover-community/mathlib4/pull/10235/commits/4b1e329abaf49f0fcf8aa6fab48096effdaf9034).

**Comment by urkud** [comment]
Is there a way to test for a function being partial? If yes, then we should do it.

**Comment by kmill** [comment]
`partial` definitions create `opaque` definitions, so it could check for that (though theoretically there could be intentional `opaque` definitions in the library, but this has not been observed yet).

Here are some checks I found:

```lean
partial def foo (n : Nat) : Nat := foo n

open Lean

def isOpaque (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci matches .opaqueInfo ..

def isUnsafe (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci.isUnsafe

def isPartial (n : Name) : MetaM Bool := do
  let ci ← getConstInfo n
  return ci.isPartial
```
With these:
* ``isOpaque `foo`` is true
* ``isUnsafe `foo`` is false (though this becomes true with the `unsafe` modifier, which also makes isOpaque become false)
* ``isPartial `foo`` is false (I think these are only used internally and no user definitions are this kind of partial)
```

File: Mathlib/GroupTheory/Perm/DomMulAct.lean (hunk 1 above)
Comment (kmill) [reply Mathlib/GroupTheory/Perm/DomMulAct.lean]: Ah, I thought it was complaining about the iota fintype instance so it looked like a bug. I didn't realize there was a `[Fintype α]` instance too. Thanks for the clarification (and for improving the lemma!)

**Comment by grunweg** [comment]
@kmill Thanks for the explanations. I had tried `isPartial`, but that did not do what I wanted (or did not go far enough?).
I've tried to check for `opaqueInfo` now; let's see if that works.

**Comment by kmill** [comment]
(My cursory understanding of isPartial is that it's for "pre definitions" for (mutually) recursive functions, letting them be added to the environment for future processing. As far as I know, this option is irrelevant to us — it has no relation to `partial` — and we'd never see it unless we're really digging around into internal details.)

**Comment by grunweg** [comment]
:bulb: That is helpful. Would you like to file PR adding this docstring? Or should I create so, so you can get help fixing it more easily :-)?

**Comment by kmill** [comment]
Feel free to create a pr and ping me and/or Joachim. He's more familiar with this part of Lean.

**Comment by grunweg** [comment]
!bench

**Comment by urkud** [comment]
!bench

**Comment by urkud** [comment]
Should we merge it now?

**Comment by grunweg** [comment]
I'm happy with merging it, and looking at post-merge benchmarking output.

**Comment by grunweg** [comment]
I am not confident there will be no regression in linting times - but as there are further such linters in the pipeline, I am not worried about that item dropping of everybody's radar.

**Comment by grunweg** [comment]
Thanks for the fixes; they look good to me.

**Comment by urkud** [comment]
!bench

**Comment by urkud** [comment]
Why is it so slow?

**Comment by grunweg** [comment]
Let me test one hypothesis: does merging all linters into one help? I just pushed a commit doing so. I benchmarking is better, I would start a discussion on zulip on how to make this linter fast(er) while preserving nice error messages.

I may not be at a computer in ~40 minutes (when CI will have finished); feel free to re-bench then.

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      if argTys[i+1:].any (·.containsFVar fv) then return none
+      return some (i, (← addMessageContextFull m!"argument {i+1} {arg} : {t}"))
+  if !(← isProp type) then
+    if let some e := declInfo.value? then
+      impossibleArgs ← lambdaTelescope e fun args e ↦ do
+        let e ← e.eraseProofs
+        return impossibleArgs.filter fun (k, _) ↦
+          k < args.size && !e.containsFVar args[k]!.fvarId!
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
```
Comment (kmill) [inline Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:52]: I saw you were having performance problems. I tried rewriting this as efficiently as I could, but unfortunately I didn't test it, so it might be worthless... Here:
```lean
/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
assumption in the type. `typesToAvoid` the list of such types.
This is the main logic underlying the linters below. -/
def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
    MetaM (Option MessageData) := do
  -- We omit inductive types and their constructors, to reduce false positives.
  -- We also omit partial declarations (via the `opaque` definitions they generate):
  -- these are less useful for theorem proving, hence the linter is less useful there.
  if declInfo.isInductive || declInfo.isCtor ||
      (declInfo matches .opaqueInfo ..) then return none
  let type := declInfo.type
  -- Early return: none of the constants to avoid appear.
  unless type.containsConst typesToAvoid.contains do
    return none
  forallTelescopeReducing type fun args ty ↦ do
    -- Compute whether the given argument is used by the body of the definition outside of proofs.
    let usedIdxs : Array Bool ← do
      if ← pure declInfo.hasValue <&&> not <$> isProp type then
        lambdaTelescope declInfo.value! fun args' e ↦ do
          let mut usedIdxs := args.map (fun _ => false)
          let e ← e.eraseProofs
          let st := collectFVars {} e
          for i in [0:args'.size] do
            usedIdxs := usedIdxs.set! i (st.fvarSet.contains args'[i]!.fvarId!)
          pure usedIdxs
      else
        pure #[]
    -- Early return: every argument is used.
    if usedIdxs.size == args.size && usedIdxs.all id then
      return none
    -- Compute an array of error messages for each superfluous argument
    let ty ← ty.eraseProofs
    let mut impossibleArgs := #[]
    let mut usedFVars : CollectFVars.State := collectFVars {} ty
    for i' in [0:args.size] do
      let i := args.size - i' - 1
      let arg := args[i]!
      let t := (← inferType arg).cleanupAnnotations
      if t.containsConst typesToAvoid.contains then
        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains t.fvarId! then
          impossibleArgs := impossibleArgs.push
            (← addMessageContextFull m!"argument {i+1} {arg} : {t}")
      usedFVars := collectFVars usedFVars t
    if impossibleArgs.isEmpty then return none
    return some <| .joinSep impossibleArgs.reverse.toList ", "
```

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
+assumption in the type. `typesToAvoid` the list of such types.
+This is the main logic underlying the linters below. -/
+def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
+    MetaM (Option MessageData) := do
+  -- We omit inductive types and their constructors, to reduce false positives.
+  -- We also omit partial declarations (via the `opaque` definitions they generate):
+  -- these are less useful for theorem proving, hence the linter is less useful there.
+  if declInfo.isInductive || declInfo.isCtor ||
+      (declInfo matches .opaqueInfo ..) then return none
```
Comment (kmill) [inline Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:31]: It's not a performance improvement, but you might like this syntax:
```lean4
if declInfo matches .inductInfo .. | .ctorInfo .. | .opaqueInfo .. then return none
```

Comment (grunweg) [reply Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:31]: I actually know this syntax - but thanks for reminding me! This started as one condition, branched out and I didn't re-visit: I will certainly use the or pattern!

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean (hunk 3 above)
Comment (kmill) [reply Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:52]: If written correctly, it should be nearly functionally identical. The only intentional difference is that the message uses `t` with cleaned-up annotations now.

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  if !(← isProp type) then
+    if let some e := declInfo.value? then
+      impossibleArgs ← lambdaTelescope e fun args e ↦ do
+        let e ← e.eraseProofs
+        return impossibleArgs.filter fun (k, _) ↦
+          k < args.size && !e.containsFVar args[k]!.fvarId!
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
+
+@[env_linter] def allOfThem : Linter where
```
Comment (urkud) [inline Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:54]:
```suggestion
/-- Temporary  hack -/
@[env_linter] def allOfThem : Linter where
```

**Comment by urkud** [comment]
!bench

**Comment by grunweg** [comment]
So it seems my theory was right: having four separate linters takes four times as long... but even the current implementation is too inefficient. Next step, try Kyle's suggestion. (I have to go bed now, but next time I work on this - or if anybody beats me to it.)

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/UnusedAssumptionInType.lean]: Thanks! I have used this version now; let's see how it fares w.r.t. speed!

File: Mathlib/Tactic/Linter/UnusedAssumptionInType.lean [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    -- Compute an array of error messages for each superfluous argument
+    let ty ← ty.eraseProofs
+    let mut impossibleArgs := #[]
+    let mut usedFVars : CollectFVars.State := collectFVars {} ty
+    for i' in [0:args.size] do
+      let i := args.size - i' - 1
+      let arg := args[i]!
+      let t := (← inferType arg).cleanupAnnotations
+      if t.containsConst typesToAvoid.contains then
+        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains t.fvarId! then
```
Comment (kmill) [inline Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:64]: Oops, this was supposed to be
```suggestion
        if usedIdxs[i]? != some true && !usedFVars.fvarSet.contains arg.fvarId! then
```

Comment (grunweg) [reply Mathlib/Tactic/Linter/UnusedAssumptionInType.lean:64]: Thanks! Let's try the new version :-)

**Comment by grunweg** [comment]
@kmill Can you make heads or tails out of [this output](https://github.com/leanprover-community/mathlib4/actions/runs/11444480102/job/31839449455)? For comparison, I'll revert one commit, to check if there are new legitimate linter errors on master.

**Comment by grunweg** [comment]
!bench

**Comment by grunweg** [comment]
To clarify: this benchmarking run tests the cumulative effects of
- merging all the linters into one (which I would really prefer not to do, for sake of user-friendliness)
- adding a first early return
I expect this to still be too slow, but let's see how much :-)

**Comment by grunweg** [comment]
Let me summarise the benchmarking results so far:
- [initial run](http://speed.lean-fro.org/mathlib4/compare/29678737-f417-40ee-bae5-eb2ec422f0a5/to/ea0e91d8-f641-406e-96eb-fa978a072335), no optimisations at all: TODO insert!
- [second run](http://speed.lean-fro.org/mathlib4/compare/29678737-f417-40ee-bae5-eb2ec422f0a5/to/ea0e91d8-f641-406e-96eb-fa978a072335): merging all linters into one, no optimisations: linting increased by +740*10⁹ instructions, wall-time by 15 seconds (in CI)
- [third run](http://speed.lean-fro.org/mathlib4/compare/4cf8ccc1-6c8f-4a8d-998a-8d741ceef6b0/to/64a43f37-fbb9-46c1-a391-0ede794462d3): merging all linters into one, no optimisations: linting increased by +150*10⁹ instructions, wall-time by ~7 seconds (in CI)

For comparison, currently, the linting step (in CI) takes 6450*10⁹ instructions or 333 seconds.

**Comment by grunweg** [comment]
!bench

**Comment by grunweg** [comment]
TL;DR @kmill Your help fixing the optimisation you proposed is welcome. I don't understand the code yet (and don't have much time this week to try to change that).

File: Mathlib/NumberTheory/JacobiSum/Basic.lean [hunk 7]
```diff
@@ -274,6 +274,7 @@ lemma exists_jacobiSum_eq_neg_one_add {n : ℕ} (hn : 2 < n) {χ ψ : MulChar F
     {μ : R} (hχ : χ ^ n = 1) (hψ : ψ ^ n = 1) (hn' : n ∣ Fintype.card F - 1)
     (hμ : IsPrimitiveRoot μ n) :
     ∃ z ∈ Algebra.adjoin ℤ {μ}, jacobiSum χ ψ = -1 + z * (μ - 1) ^ 2 := by
+  classical
```
Comment (Command-Master) [inline Mathlib/NumberTheory/JacobiSum/Basic.lean:277]: What is the reason for this change?

**Comment by urkud** [comment]
Note: I'm working on `HasFDerivAt.pi` etc. I want to do it the right way (drop the `Normed*` assumptions at the same time), should be ready for review soon.

**Comment by Vierkantor** [comment]
This PR is flagged as awaiting-author, what are you planning to do before this is ready for review? I would like to see this PR in Mathlib somewhat soon so I am willing to help out fix some of the build errors.

**Comment by grunweg** [comment]
Can this PR be closed now, or is there anything else that's missing? If so, we could track that in #35340.
